---
title:  "CRISPR analysis report"
#author: "Martin Demko"
#date:   "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    toc: true
    toc_depth: 2
    toc_float: 
      collapsed: no
      smooth_scroll: yes
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(ggplot2)
library(plotly)
```

#### *Project: `r snakemake@params$name`*
#### *Date: `r format(Sys.time(), '%d %B, %Y')`*

<!--
A Way how to use config in Rmd file (assuming we have cfg in params in snakemake rule):
`r rjson::fromJSON(snakemake@params$cfg)$macs_padj_filter[1]`
-->

## Description

CRISPR analysis pipeline is based on ...

## Pipeline steps

* Step 1: ...

* Step 2: ...

## Pipeline outputs

Pipeline directory content description (only important files/folders are described). The pipeline starts with the raw fastq files ...

* `preprocessed_data` (folder) - contains files with preprocessed sequences and QC data of clean fastq files as a result of [FastQC tool](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/).
    * `*_clean_SE.fastq.gz` (compressed file) - is a clean fastq file as a result of trimming process by [Cutadapt tool](https://cutadapt.readthedocs.io/en/stable/).
    * `*.inserts.fa` (file) - contains unique inserts (`r snakemake@params$guide_len`nt long pieces of reads). This is input for following alignment process.
    * `*.inserts.counts` (file) - contains abundance of unique inserts in original fastq file.
```{r echo = FALSE, results = "asis"}
path <- "preprocessed_data"
for (i in list.files(paste0(snakemake@params$wdir, path), "fastqc.html")){
  cat(sprintf("\t* [%s](%s/%s)\n", basename(i), path, i))
}
```

```{r stat1, echo = FALSE, results = "asis"}
stat1 <- fread(paste0(snakemake@params$wdir, "mapped/all_samples.stats.tsv"))
stat1[,sample:=gsub(".stats.tsv", "", sample)
setorder(stat1, sample)
mstat1<-melt(stat1,id.vars = "sample")

total_inserts <- mstat1[variable %like% "original", ]$value[1]
mstat1[variable %like% "targeted inserts", `:=`(perc= round(value/total_inserts*100,2), perc_pos=value/2)]

stat1_read <- ggplot(mstat1[variable %like% "reads"],aes(value, sample, fill=variable)) +
geom_col(position = "dodge") +
theme_bw() +
theme(legend.position = "bottom") +
xlab("# of reads") + ylab("") +
scale_fill_manual(name="", values = c("firebrick","royalblue"), labels = c("mapped","unmapped"))

stat1_inserts <- ggplot(mstat1[variable %like% "targeted inserts"],aes(value, sample, fill=variable)) +
geom_col(position = "dodge") +
geom_text(aes(x=perc_pos, label = paste0(perc, "%")), position = position_dodge(width = 1), hjust = -0.1) +
theme_bw() +
theme(legend.position = "bottom") +
xlab("# of inserts") + ylab("") +
scale_fill_manual(name="", values = c("firebrick","royalblue"), labels = c("targeted","not targeted"))

```

* `mapped` (folder) - contains raw results of alignment of inserts to reference (`r snakemake@params$ref`) by [HS-BLASTN](https://github.com/chenying2016/queries/tree/master/hs-blastn-src) along with a summary table
  * [all_samples.stats.tsv](./mapped/all_samples.stats.tsv) (file) - summary table from alignment
  * `*.hsblastn.tsv` (file) - is a tabular result of alignment of particular sample's inserts to the reference

```{r mapped-read, echo = FALSE, results = "asis"}
ggplotly(stat1_read)
```

```{r mapped-inserts, echo = FALSE, results = "asis"}
ggplotly(stat1_inserts)
```

<!--
```{r echo = FALSE, results = "asis"}
path = "mapped"
for (i in list.files(paste0(snakemake@params$wdir, path), "hsblastn.tsv", recursive=T)){
  cat(sprintf("\t* [%s](%s/%s)\n", basename(i), path, i))
}
```
-->

```{r stat2, echo = FALSE, results = "asis"}
stat2 <- fread(paste0(snakemake@params$wdir, "counts/all_samples_report.tsv"))
mstat2 <- melt(stat2,id.vars = c("sgRNA", "Gene"), variable.name = "sample")
mstat3 <- mstat2[, value_gene:=sum(value), by=.(Gene,sample)]
mstat3[, var_per_gene:=var(value_gene), by=.(Gene)] # get the variance per gene
setorder(mstat3, -var_per_gene,Gene, sample)

top10_genes <- head(unique(mstat2[!Gene %like% "Control",]$Gene),10)
stat2_top10 <- stat2[Gene %in% top10_genes]
```


* `counts` (folder) - contains results with inserts counts.
  * [all_samples_report.tsv](./counts/all_samples_report.tsv) (file) - raw counts for sample data against reference guide sequences
  * [all_samples_report.pdf](./counts/all_samples_report.pdf) (file) - report summarising alignment data with various statistics
  * `*_graphs.pdf` (file) - shows inserts histograms:
```{r echo = FALSE, results = "asis"}
path <- "counts"
for (i in list.files(paste0(snakemake@params$wdir, path), "graphs.pdf", recursive=T)){
  cat(sprintf("\t* [%s](%s/%s)\n", basename(i), path, i))
}
```

* `DE_results` (folder) - contains results of differential expression
  * `edgeR` (folder) - contains results processed by [edgeR tool](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
    * `*.gene_summary.tsv` (file) - result of DE on gene level:
```{r echo = FALSE, results = "asis"}
path <- "DE_results/edgeR"
for (i in list.files(paste0(snakemake@params$wdir, path), "gene_summary.tsv", recursive=T)){
  cat(sprintf("\t* [%s](%s/%s)\n", basename(i), path, i))
}
```
    * `*.sgrna_summary.tsv` (file) - result of DE on sgRNA level:
```{r echo = FALSE, results = "asis"}
path <- "DE_results/edgeR"
for (i in list.files(paste0(snakemake@params$wdir, path), "sgrna_summary.tsv", recursive=T)){
  cat(sprintf("\t* [%s](%s/%s)\n", basename(i), path, i))
}
```
    * `*.graphs.pdf` (file) - summary results with plots for TOP `r snakemake@params$top` genes:
```{r echo = FALSE, results = "asis"}
path <- "DE_results/edgeR"
for (i in list.files(paste0(snakemake@params$wdir, path), "graphs.pdf", recursive=T)){
  cat(sprintf("\t* [%s](%s/%s)\n", basename(i), path, i))
}
```
  * `MAGeCK` (folder) - contains results processed by [MAGeCK tool](https://bioconductor.org/packages/release/bioc/html/edgeR.html) tool
    * `*.gene_summary.txt` (file) - result of DE on gene level:
```{r echo = FALSE, results = "asis"}
path <- "DE_results/MAGeCK"
for (i in list.files(paste0(snakemake@params$wdir, path), "gene_summary.tsv", recursive=T)){
  cat(sprintf("\t\t* [%s](%s/%s)\n", basename(i), path, i))
}
```
    * `*.sgrna_summary.txt` (file) - result of DE on sgRNA level:
```{r echo = FALSE, results = "asis"}
path <- "DE_results/MAGeCK"
for (i in list.files(paste0(snakemake@params$wdir, path), "sgrna_summary.tsv", recursive=T)){
  cat(sprintf("\t\t* [%s](%s/%s)\n", basename(i), path, i))
}
```
    * `*.graphs.pdf` (file) - summary results with plots for TOP `r snakemake@params$top` genes:
```{r echo = FALSE, results = "asis"}
path <- "DE_results/MAGeCK"
for (i in list.files(paste0(snakemake@params$wdir, path), "graphs.pdf", recursive=T)){
  cat(sprintf("\t\t* [%s](%s/%s)\n", basename(i), path, i))
}
```

* [`final_report.html`](./final_report.html) (file) - this report.
